pyproject.toml

```diff
@@ -6,35 +6,32 @@
 # the most recent version on PyPI:
 #
 #     "pybind11>=2.4.3,<2.5.0",
+#
+# Upper bounds in release branches must have notes on why they are added.
+# Distro packages can ignore upper bounds added only to prevent future
+# breakage; if we add pins or bounds because of known problems then they need
+# them too.
 
 [build-system]
 build-backend = 'mesonpy'
 requires = [
-    "meson-python>=0.15.0",
-    "Cython>=3.0.8",  # when updating version, also update check in meson.build
+    # The upper bound on meson-python is pre-emptive only (looser on purpose,
+    # since change of breakage in 0.16/0.17 is low)
+    "meson-python>=0.15.0,<0.18.0",
+    # The upper bound on Cython is pre-emptive only
+    "Cython>=3.0.8,<3.1.0",  # when updating version, also update check in meson.build
+    # TODO: TBD, depending on https://github.com/pybind/pybind11/issues/5009
```

Comment:
Looks like we will need this resolved before numpy 2.0.0rc1, xref https://github.com/numpy/numpy/issues/26010. It's not blocking for us to release 1.13.0rc1 though, since things build. But we'd like to bump to a newer pybind11 release with the proper fix.